include/envoy/api/api.h

```diff
@@ -5,7 +5,6 @@
 
 #include "envoy/common/time.h"
 #include "envoy/event/dispatcher.h"
-#include "envoy/event/timer.h"
```

Comment:
I assume the change to remove `createScheduler(Libevent::BasePtr&)` from the public interface is going to come in a hopefully small follow up? (Just making sure I understand the plan.)

Reply:
I don't have a definitive plan beyond this actually. TimeSystem needs to exist and be libevent aware to enable the simulated-time system implementation.

What this accomplishes is making all of prod other than dispatcher be aware of that, but I found in the course of this PR there are other modules that need to know deeply about libevent.

So a holistic strategy (which this PR makes easier to implement) is to factor out libevent completely.

I think that needs to include also the iosystem work and other FD-abstraction stuff. We should probably try to get a slack-channel or something going on to sort all this out as a group.

Reply:
But why is the `createScheduler()` API needed at all at this point? Isn't it an internal implementation detail that can be completely hidden from the public interface?

Reply:
Maybe we can chat a little later via slack? createScheduler() is the hook that SimulatedTimeSystem overrides so that timers work properly against test-controlled time. Happy to brainstorm though.

Reply:
Sure let's chat later. I think you can have a `SimulatedTimeDispatcher` which just overrides `createTimer()` and that should work very well.

Reply:
I think that may be possible but it would be a pretty large change from the current condition, and I'm not seeing how that improves things. In particular, there's still a significant amount of code outside of dispatcher_impl.cc that knows a lot about libevent, so we're leaking this abstraction independent of TimeSystem.

Here's why I like TimeSystem as an abstraction:

Right now a test-writer can decide what time-system to use per-test basis by just declaring a SimulatedTimeSystem or RealTimeSystem in the test-class, and passing it into constructors, etc (or relying on my TestGlobal<Class> hack, to allow 0-constructor-arg mocks). In particular, if you ask for a SimulatedTimeSystem you can then directly manipulate time with it, and it will work properly with the timers created via any dispatcher. You can also use simulated sleep, etc.

The TimeSystem abstraction provides a single point of control for how we are working with time in tests, affecting the way we query time, sleep, register timers, etc. I think there are a lot of tests that need to know about time that don't need to know about dispatchers, so I'd rather not tie the time-semantics into a Dispatcher subclass. Also Dispatcher has other subclasses, which make that plan a bit messy IMO, though we can probably hammer that out with multiple inheritance.

It's unfortunate that TimeSystem exposes a libevent structure in its API but I feel like that problem would best be solved by having an abstract base class over libevent, that might possibly allow substituting an alternate event manager. In the meantime at least this PR limits the blast-radius of that libevent exposure.

Reply:
> In particular, there's still a significant amount of code outside of dispatcher_impl.cc that knows a lot about libevent, so we're leaking this abstraction independent of TimeSystem.

Only within /source/common/event though. Nothing outside of that knows about libevent, nor should it in my opinion. That's my main problem here. I realize we are entering bikeshedding territory here, but can't this interface be private to /source/common/event and any derived classes? That is really what I'm after. I think that fact that this is exposed in a public timer API is a large anti-pattern. I'm fairly positive this is possible without a terribly large amount of work?

Reply:
Basically what I'm suggesting it to have a `SimulatedTimeApi` which produces a `SimulatedTimeDispatcher` concrete object, and then tests that use simulated time with start with that type of API object.

Reply:
@jmarantz and I discussed offline and I think we are largely on the same page now, so we can ignore this comment thread for this PR and continue the discussion in subsequent PRs.